### ISSUE TITLE
Liquid Status

### ISSUE THREAD

**Issue by Harrith** [TrinityCore/TrinityCore#22020]
**Description:**

Server is incorrectly determining the liquid status.

**Current behaviour:**

In some locations, server incorrectly determines the liquid status. Somewhere player can breathe underwater and somewhere player cannot breathe when he is not underwater.

Example of locations where player can breathe underwater:  
.go xyz -4634.93 3539.738 -114.65 0

Example of locations where player should be able to breathe:
.go xyz -6384 4078 -431 0     - Nespirah, here is also Fatigue and should not be
.go xyz -9760 -944 57 1

EDIT: with map version 1.7 some locations were fixed, so these locations were removed from example list

**Expected behaviour:**

Player should/should not be able to breathe it depends on if he is/is not underwater.

**Steps to reproduce the problem:**

Just use some of commands above to teleport to the location.

**Branch(es):**

4.3.4

**TC rev. hash/commit:** 
ad3dff69c122

**Operating system:** Windows 10

**Comment by Killyana** [comment]
4.3.4 is no longer supported, but this issue is valid for master.

**Comment by Ovahlord** [comment]
I just did a few tests on my fork that includes every master mmap / vmap changes and the following coordinates are working as intended:
.go xyz -4640 3537 -122 0:
![wowscrnshot_061318_004149](https://user-images.githubusercontent.com/18347559/41321076-9685f82c-6ea2-11e8-8611-d3a348f8cba9.jpg)
.go xyz -6075 4266 -360 0
![wowscrnshot_061318_004215](https://user-images.githubusercontent.com/18347559/41321086-a276bb58-6ea2-11e8-9a39-544e6300a091.jpg)
.go xyz 2069 -4495 77 1
![wowscrnshot_061318_003911](https://user-images.githubusercontent.com/18347559/41321088-a8822c8a-6ea2-11e8-919c-63148cd6b6f0.jpg)

.go xyz -6496 -3323 -87 1
![wowscrnshot_061318_003945](https://user-images.githubusercontent.com/18347559/41321095-b1d7daf0-6ea2-11e8-83ad-480b553abd40.jpg)


Nespira in Vashjir and The Vinal Dam however indeed have a breath bar issue but no fatigue:
![wowscrnshot_061318_004053](https://user-images.githubusercontent.com/18347559/41321104-c2e98492-6ea2-11e8-9b05-0b63e6056171.jpg)
![wowscrnshot_061318_004113](https://user-images.githubusercontent.com/18347559/41321110-c63fefe6-6ea2-11e8-97f3-f4551c1ae627.jpg)

**Comment by Killyana** [comment]
The 2 last ones are not, as there's no water.

**Comment by Harrith** [comment]
I am sorry I used fork from Aokromes and there, for some reason, is missing commit e8586b0bcba13416f0b, so there are maps version 1.6 and there this bug occurs and maps in your Ovahlord fork are in version 1.7.
But unfortunately in these maps this bug also occurs too but in fewer cases, like here:
.go xyz -4634.93 3539.738 -114.65 0

**Comment by Ovahlord** [comment]
.go xyz -4634.93 3539.738 -114.65 0 works for me (tm)
It appears that the only common issue between master and 434 is the breath issue in Nespirah in Vashjir and the Vinal Dam in Uldum
![wowscrnshot_061318_104500](https://user-images.githubusercontent.com/18347559/41340119-e85d9d54-6ef6-11e8-8e18-5f671e7a8672.jpg)

**Comment by Harrith** [comment]
That's wierd look at the minimap. I think we are not in the same position
![breathtimer](https://user-images.githubusercontent.com/40215976/41340537-e8577342-6ef7-11e8-9112-336c075c1d4d.png)

**Comment by Ovahlord** [comment]
I was like moving up a bit to test if they is any difference when swimming arround.

**Comment by Harrith** [comment]
I made some changes but I am not sure if they are valid:
```
--- a/src/common/Collision/Management/VMapManager2.cpp
+++ b/src/common/Collision/Management/VMapManager2.cpp
@@ -295,7 +295,7 @@ namespace VMAP
                     ASSERT(info.hitModel);
                     type = info.hitModel->GetLiquidType();  // entry from LiquidType.dbc
                     if (reqLiquidType && !(GetLiquidFlagsPtr(type) & reqLiquidType))
-                        return false;
+                        return true;
                     ASSERT(info.hitInstance);
                     if (info.hitInstance->GetLiquidLevel(pos, info, level))
                         return true;
```

```
--- a/src/server/game/Maps/Map.cpp
+++ b/src/server/game/Maps/Map.cpp
@@ -2585,6 +2584,10 @@ ZLiquidStatus Map::GetLiquidStatus(PhaseShift const& phaseShift, float x, float
    if (vmgr->GetLiquidLevel(terrainMapId, x, y, z, ReqLiquidType, liquid_level, ground_level, liquid_type))
    {
        TC_LOG_DEBUG("maps", "GetLiquidStatus(): vmap liquid level: %f ground: %f type: %u", liquid_level, ground_level, liquid_type);
+
+       if (liquid_level == INVALID_HEIGHT)
+          return LIQUID_MAP_NO_WATER;
+
        // Check water level and ground level
        if (liquid_level > ground_level && z > ground_level - 2)
        {
```

Maybe GetLiquidLevel should return enum of 3 values instead of boolean?

Unfortunately there are still problems in some places like here:
.go xyz -4610 3984 -70.615 0    (here should not be breath timer)
.go xyz -9798 -869 94 1             (here should be breath timer)

**Comment by Shauren** [comment]
You should try getting the liquid level logic from 3.3.5 branch, I did a bunch of improvements there that are not in master yet

**Comment by Harrith** [comment]
Ok thanks, I will try it.

**Comment by Harrith** [comment]
Unfortunately I have found place were breath timer is wrong. There is no water and player has breath timer. It seems like it's map issue so I don't know if I should open another issue because it's not really connected to the origin issue.
I opened the maptile in Recast and it looks like this.

![dragonblight1](https://user-images.githubusercontent.com/40215976/43915390-dead58da-9c0a-11e8-997e-7a3f7d1763ed.png)

![dragonblight2](https://user-images.githubusercontent.com/40215976/43915426-f3d85d22-9c0a-11e8-981f-20b8675e5283.png)

Maybe it would be better if you look at the map yourself, I am not sure if these pictures are enough to describe the problem.

.go xyz 3697.51 933.39 54 571

**Comment by Shauren** [comment]
I know about that, #20221

**Comment by Harrith** [comment]
Ok, sorry I haven't noticed that issue.

**Comment by Mesielecat** [comment]
Exist another issue related with the  Liquid Status in Vashjir, after the last updates of masp in master, Vashjir is working really well, first you going to have breath bar in vashjir until you recieve the spell Sea Legs 73701, triggered by Sea Legs 76377. Until this behaviour all is working well in the map, but the spell 73701 have AuraInterruptFlag Underwater, so this going to be removed after you leave from the water, then if yoo walk in the ground they are some place where:

``    
// Check water level and ground level
``
``
    if (liquid_level < ground_level || z < ground_level)
``

the difference between z and ground level is around 0.001f but enough for cancel the effect and return the status of LIQUID_MAP_NO_WATER, this could be handle with a tolerance in the map extractor. This was the only issue that I detect in Vashjir

**Comment by Faq** [comment]
@Mesielecat your comment would be complete with example coords.

**Comment by Mesielecat** [comment]
> @Mesielecat your comment would be complete with example coords.

Yeah sure I can share coordinates

**Comment by acidmanifesto** [comment]
https://github.com/TrinityCore/TrinityCore/issues/27946

**Comment by Mesielecat** [comment]
I starting moving from .go xyz -4705.07, 3616.67, -120.287,  0  to  .go xyz -4669.68, 3565.97, -140.87, 0 
always walking near to the ground and when you end the path return walking near to the ground, you going to see my explanation before. Remember have aplicated the aura 73701 Sea Legs, this aura as I said before is removed by underwater aurainterruptflag

**Comment by Mesielecat** [comment]
> #27946

confirmed this coordinate too

**Comment by Mesielecat** [comment]
> #27946

Research for that issue:

in this coordinate Desolace Area 6514 Ex: X: -1464.8, Y: 2676.61, Z: 76.8318, MapId: 1 the WMO have liquid level of 80.84f, that is the reason for underwater status.

**Comment by Mesielecat** [comment]
Reviewing this coordinate .go xyz -9760 -944 57 1 can be fixed increased the GROUND_HEIGHT_TOLERANCE 0.625f
